cpp/src/arrow/compute/kernels/vector_replace.cc

```diff
@@ -442,23 +442,409 @@ struct ReplaceWithMaskFunctor {
+    *output = array;
+    return Status::OK();
+  }
+};
+
+template <typename Type>
+struct FillNullForwardFunctor {
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    switch (batch[0].kind()) {
+      case Datum::ARRAY: {
+        auto array_input = *batch[0].array();
+        int64_t last_valid_value_offset = -1;
+        return FillNullForwardArray(ctx, array_input, out, array_input,
+                                    &last_valid_value_offset);
+      }
+      case Datum::CHUNKED_ARRAY: {
+        return FillNullForwardChunkedArray(ctx, batch[0].chunked_array(), out);
+      }
+      default:
+        break;
+    }
+    return Status::NotImplemented(
+        "Unsupported types for drop_null operation: "
+        "values=",
+        batch[0].ToString());
+  }
+
+  static Status FillNullForwardArray(KernelContext* ctx, ArrayData& array, Datum* out,
+                                     ArrayData last_valid_value_chunk,
+                                     int64_t* last_valid_value_offset) {
+    ArrayData* output = out->array().get();
+    /*if (!output->buffers[0]) {
+      ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(array.length));
+    }*/
+    output->length = array.length;
+    int8_t direction = 1;
+
+    if (array.MayHaveNulls()) {
+      return FillNullExecutor<Type>::ExecFillNull(
+          ctx, array, array.buffers[0]->data(), output, direction, last_valid_value_chunk,
+          last_valid_value_offset);
+    } else {
+      if (array.length > 0) {
+        *last_valid_value_offset = LastElementOffset(array, direction);
+      }
+      *output = array;
+    }
+    return Status::OK();
+  }
+
+  static Status FillNullForwardChunkedArray(KernelContext* ctx,
+                                            const std::shared_ptr<ChunkedArray>& values,
+                                            Datum* out) {
+    if (values->null_count() == 0) {
+      *out = Datum(values);
+      return Status::OK();
+    }
+    if (values->null_count() == values->length()) {
+      *out = Datum(values);
+      return Status::OK();
+    }
+    std::vector<std::shared_ptr<Array>> new_chunks;
+
+    if (values->length() > 0) {
+      ArrayData array_with_current = *values->chunk(/*first_chunk=*/0)->data();
+      int64_t last_valid_value_offset = -1;
+      for (const auto& chunk : values->chunks()) {
+        auto buffer_size = chunk->length() * bit_width(values->type()->id()) / 8;
+
+        std::unique_ptr<ArrayBuilder> builder;
+        RETURN_NOT_OK(MakeBuilder(ctx->memory_pool(), values->type(), &builder));
+        RETURN_NOT_OK(builder->Reserve(chunk->length()));
+        ARROW_ASSIGN_OR_RAISE(auto array_output, builder->Finish());
+        ARROW_ASSIGN_OR_RAISE(array_output->data()->buffers[1],
+                              ctx->Allocate(buffer_size));
+        ARROW_ASSIGN_OR_RAISE(array_output->data()->buffers[0],
+                              ctx->AllocateBitmap(chunk->length()));
```

Comment:
Instead of doing this, can we push the allocation into the kernel implementation itself?

It would have to be something like
```cpp
auto* output = out->mutable_array();
auto bit_width = checked_cast<const FixedWidthType&>(*output.type).bit_width();
auto data_bytes = bit_util::BytesForBits(bit_width * chunk->length);
ARROW_ASSIGN_OR_RAISE(output->buffers[0], ctx->AllocateBitmap(chunk->length));
ARROW_ASSIGN_OR_RAISE(output->buffers[1], ctx->Allocate(data_bytes));
```

Reply:
The reason why I prefer not to do this here is because in the case of strings, it is redundant with what the kernel itself does.

As mentioned previously, the kernel registration below will have to be adjusted:
```
kernel.mem_allocation = MemAllocation::type::COMPUTED_NO_PREALLOCATE;
```

Reply:
Alternatively, you can try a check like the following right here:

```cpp
if (is_fixed_width(out->type()->id())) {
   // ...snippet from above
}
// otherwise, the kernel implementation itself will allocate
```

But we shouldn't need a builder just to allocate the output.

Reply:
LGTM, thanks.